### ISSUE TITLE
Universal Kedro Deployment (Part 4) - Embedding kedro pipelines in third-party applications 

### ISSUE THREAD

**Issue by Galileo-Galilei** [kedro-org/kedro#3540]
# Introduction

### Overview

:memo: **Document type** The document is mostly a "User Research" which focuses *what* functional needs should be added to kedro, but it also sometimes slips to a "Design document" which suggests *how* they could be integrated to the core kedro framework. The 2 parts are clearly separated, because I think the user research (mainly a compilation of existing kedro issues) is worth sharing, even if we do not end up with the proposed implementation.

:busts_in_silhouette: **Target Audience** : Kedro core team / plugin developers / Mlops engineer who puts kedro in production. I try to keep the issue as self-contained as possible, but I still assume the reader know the default kedro objects (runner,pipeline, catalog...), and how ``KedroSession.run`` works under the hood. 

:pray: **Credits** : The mind behind part / most of the design and the thoughts described hereafter is @takikadiri. I mostly reformulate, clarify and try to give a comprehensive overview of the issues we are trying to solve and how we solve them. 

:books: **Additional resources**: Most of the features described hereafter are implemented in the [kedro-boot plugin](https://github.com/takikadiri/kedro-boot), however the technical implementation sometimes differ for subtle reasons. You can find examples on how to use it for the features described in the "User research" section in the  [kedro-boot examples repo](https://github.com/takikadiri/kedro-boot-examples)

:bangbang: **Important note** : The ``kedro-boot`` plugin also provides other features, especially to launch an app from a kedro entrypoint (called [standalone mode](https://github.com/kedro-org/kedro/issues/2169#issuecomment-1722914953) in this comment). This is out of scope of this issue. I' 

### TL;DR

We need to make the session being runnable multiple times, and optimize latency to serve  lots of uses cases (serving, dynamic pipelines, ...). In summary, we should make below pseudo code possible : 

```
session=KedroSession.create(project_path)
session.preload(artifacts=["my_model", "my_encoder"]) # :sparkles: FEATURE 1: preload artifacts, cache them and not not release them between runs for speed
session.run(pipeline="my_pipeline", runtime_data={"data": data1}) # :sparkles: FEATURE 2: inject data at runtime from the session
session.run(pipeline="my_pipeline", runtime_params= {"customer_id": id}) # :sparkles: FEATURE 3: run the same session mulitple times + :sparkles: FEATURE 4: inject runtime params at... runtime  (as the name says!) instead of instantation time 
session.release_cache(artifacts=["my_model", "my_encoder"]) # free memory when needed
```

There are a bunch of technical optimisations (for speed, ``kedro-viz`` compatibility and ability to inject parameters truly at runtime) which are needed under the hood. 

# User research : embedded deployment patterns

### Functional need 1 : Triggering kedro from third party applications which owns the entrypoint

There is a very common "deployment pattern" which consists to run the ``KedroSession`` programmatically in another python program. It consists in running (more or less) the following snippet: 

```python
# SNIPPET 1 : naive code
from kedro.framework.project import bootstrap_project
from kedro.framework.session import KedroSession

session=KedroSession.create(project_path)
session.run(pipeline="my_pipeline")
```

Overall, this is well described by the following issue: https://github.com/kedro-org/kedro/issues/2169

Multiple use cases are well identified: 
1. Expose the pipeline from an API to trigger the run (see https://github.com/kedro-org/kedro/issues/3094, https://github.com/kedro-org/kedro/issues/2182, https://github.com/kedro-org/kedro/discussions/2134, https://github.com/kedro-org/kedro/issues/1846, https://github.com/kedro-org/kedro/discussions/2134)

```python
from fastapi import FastAPI
app = FastAPI(title="Spaceflights shuttle price prediction")

@app.post("/predict")
def predict_data():
	session=KedroSession.create(project_path)
	session.run(pipeline="my_pipeline")
	
	return data.to_dict()
```

2. Trigger the pipeline from a frontend low-code tool (streamlit, dash, shiny...)(https://github.com/kedro-org/kedro/issues/115, https://github.com/kedro-org/kedro/issues/2169). A code example can be found here: https://github.com/takikadiri/kedro-boot-examples/blob/29438e45adb581b060f43c953c2557badeafd7d1/src/spaceflights_streamlit/app.py#L1-L33
3. Launch kedro from Datarobot : https://github.com/kedro-org/kedro/issues/2169#issue-1517295636
4. The "type 1" dynamic pipeline as defined in https://github.com/kedro-org/kedro/issues/2627#issue-1736477844 : people try to run the same pipeline multiple times with slightly different inputs (list of countries, hyperparamters,, several days for a time serie... someone even mentioned a monte carlo simulation here: https://linen-slack.kedro.org/t/14162145/hi-there-what-s-the-best-way-to-run-a-monte-carlo-simulation#48ef7630-854f-4e98-b698-3534f80a05b7)
5. Eventually in jupyter notebook, as described in https://github.com/kedro-org/kedro/issues/1273#issue-1146069665
6. We can eventually imagine people deploying kedro pipelines by running pipelines in airflow tasks "programatically" (with ``session.run()``) instead of trying to "map" kedro objects (e.g. nodes or namespaced pipelines) to airflow tasks automatically. 

:white_check_mark: Above "naive" code is valid in ``kedro>=0.18``. So... is that already ok? Clearly not, because of the next paragraph. 

### Functional need 2 : Passing data at runtime and getting the results

This long-demanded feature is described in details in https://github.com/kedro-org/kedro/issues/2169 so I'll try not to duplicate it here. The point is that almost all use cases described in previous paragraph need to inject "some data" at runtime, e.g. : 
- loops (hyperparameters tuning, montecarlo simulation)
- api injection

and retrieve the results to in memory, e.g. do something like this: 

```python
metrics=[]
for params in params_grid:
	computed_metric=session.run(pipeline="my_pipeline", runtime_params=params)
	metrics.append(computed_metric)
...
do_something_with_computed data(metrics)
```

Without the ability to pass data at runtime, all use cases presented above are not feasible so ``SNIPPET 1`` is hardly useful. We need to find a way to circumvent the current limitations to pass data at runtime. 

#### Injecting data & parameters

This is what is covered https://github.com/kedro-org/kedro/issues/2169.

```python
# SNIPPET 2 :  naive pseudo-code with runtime data injection
from kedro.framework.project import bootstrap_project
from kedro.framework.session import KedroSession

data = get_data()# where data is anything, retrieved from the context: the API input in case 1, the frontend in case 2, some parameters  in case 4, literally anything computed on the fly in case 5 ...)
session=KedroSession.create(project_path)
session.run(pipeline="my_pipeline", runtime_data= {"my_catalog_dataset": data}) # this runtime_data key does not exist, this is the suggestion of #2169
```

The current workaround consists in rewriting the ``KedroSession.run`` method in your app with many private methods which causes a lot of maitenance issues, because it becomes really hard to upgrade yout kedro version. We'll cover it in a next paragraph.

#### Injecting globals & runtime_params

A couple of issues suggest that users do not want to override the full data but only pass some ``globals`` or ``runtime_params`` to be resolved in the catalog : https://github.com/kedro-org/kedro/issues/1723. 
The typical uses cases consist in using the catalog to : 
- trigger a parametrized SQL request (with a date, a customer_id...)
- call a REST API endpoint with API dataset 
- change a filepath where the result is read/saved dynamically

```python
# SNIPPET 3 :  naive pseudo-code with runtime data injection
from kedro.framework.project import bootstrap_project
from kedro.framework.session import KedroSession

session=KedroSession.create(project_path)
session.run(pipeline="my_pipeline", runtime_globals= {"my_global": time.time()}) # this ``runtime_globals`` can be passed as ``runtime_params``
```

There are 2 *big* problems in the current implementation :
- the ``extra_params`` (*note that name is inconsistent, it is called ``params`` in the CLI, ``extra_params`` in the session and ``runtime_params`` for the resolver, I strongly suggest we normalize the name which is confusing*) key exist, but is currently a [``KedroSession.create`` argument](https://github.com/kedro-org/kedro/blob/2e64459a021bd22d79bd322d9bb87ea22f30c5f2/kedro/framework/session/session.py#L125-L132) instead of
[``KedroSession.run`` arguments](https://github.com/kedro-org/kedro/blob/2e64459a021bd22d79bd322d9bb87ea22f30c5f2/kedro/framework/session/session.py#L273-L285).  This means that we cannot override these extra params on each run. This tightly couples the 2 methods ``create`` and ``run``. It currently makes sense because [Kedro makes the strong assumption that 1 session = 1 run](https://github.com/kedro-org/kedro/pull/1329) and even raises an exception (https://github.com/kedro-org/kedro/blob/2e64459a021bd22d79bd322d9bb87ea22f30c5f2/kedro/framework/session/session.py#L324-L329) if a user attemps to run it multiple times, but we can see that it is a strong limitation that is discussed below. 
- the ``KedroSession.run`` method access the catalog through ``context.catalog`` attribute. However, the catalog is already resolved at this step and it is no longer possible to inject globals without rebuilding the entire catalog manually (inlcuding rewriting all the load/merge logic between environments ...). The issue #2973 describes in details what the problem is and suggests a potential a solution. There is a more general issue about decoupling loading and resolving configuration https://github.com/kedro-org/kedro/issues/2481. ``kedro-boot`` attempts to solve this issue by enabling a new type of parametrize query in the catalog with the ``[[ ]]`` syntax instead of the jinja ```${}` syntax, but this feels a very bad workaround not sustainable on the long run. We'll do better [with a custom resover in a next version](https://github.com/takikadiri/kedro-boot/issues/7).

### Technical requirements 1 : Speed of execution

Above "business" use cases (especially API serving) requires speed of execution, hence any overhead induced by kedro negatively impact their feasibility. A couple of seconds of overhead is acceptable on a 5 mn batch, but less on an API serving preidction swith high latency (say <100ms).
There are at least 3 known performance issues of the ``KedroSession``: 
- **CLI issues due to click** plugins collection (https://github.com/kedro-org/kedro/issues/2879). This is not relevant here since we focus on programmatical use. 
- Each **session run calls ``session.load_context()`` which is slow**. The goal is to provide the save and load version by using ``context._get_catalog(...)`` , as described in https://github.com/kedro-org/kedro/blob/2e64459a021bd22d79bd322d9bb87ea22f30c5f2/kedro/framework/session/session.py#L334. The aforementioned issue makes clear why it is done as is currently, but this feels a high cost for very low benefit. If we can avoid calling ``load_context``, it will speed up the session run. 
- **Artifacts preloading** : kedro loads data lazily in the first nodes which needs it, and release it once the last nodes which needs it has finished compute. This is the expected behaviour in batch mode, but in API mode you will run the pipeline multiple times, and you don't want the I/O to become the performance bottleneck. It would be useful to enable to force loading some datasets (e.g., a ML model) in memory *before* running the session, and do not release them until being explictly told. ``Mlflow`` call such datasets "artifacts", see https://github.com/Galileo-Galilei/kedro-mlflow/blob/e0033c5072c929a4c26cfaeaf61fcedf93d36522/kedro_mlflow/mlflow/kedro_pipeline_model.py#L174-L219. 

```python
# SNIPPET 4 :  naive pseudo-code with artifact preloading

session=KedroSession.create(project_path)
session.preload(artifacts=["my_model", "my_encoder"]) # syntax TBD
session.run(pipeline="my_pipeline", runtime_data= {"data": data1}) # the catalog.load("my_model") should use the preloaded cached data. Maybe reuse CachedDataSet? 
session.run(pipeline="my_pipeline", runtime_data= {"data": data2}) # the catalog.load("my_model") should use the preloaded cached data. Maybe reuse CachedDataSet? 
session.release_cache(artifacts=["my_model", "my_encoder"]) # syntax TBD
```

### Technical sub-requirements 2 : Running multiple times a single session

If we summarize the last paragraphs, it boils down that we need to make the session runnable mulitples times without the need to create it for each run which creates a tight coupling between the 2 methods. We show that it would enable: 
- Pass data dynamically  to ``Kedrosession.run`` 
- Move ``extra_params`` to ``Kedrosession.run`` instead of ``KedroSession.create``
- Pass ``runtime_params`` (or globals) to ``Kedrosession.run`` and resolve the catalog on each run instead on Session instantiation
- Preload some datasets after create and do not release them at the end of each run to increase runtime speed of execution

However, there are **deep investigations** to make before enabling this, because kedro makes the assumption that [1 session = 1 run](https://github.com/kedro-org/kedro/pull/1329) on purpose. According to the issue, this is done to simplify kedro-viz experiment tracking functionality. Changing this assumption **may be impossible to be backward compatible with existing experiment tracking in kedro-viz**, as described in https://github.com/kedro-org/kedro/issues/1273. However this potential breaking change should be thought in regards of https://github.com/kedro-org/kedro-viz/issues/1624, which suggests low adoption of the experiment tracking functionality, and my feeling is that enabling running multiple sessions is worth breaking experiment tracking, but I am overly biased here :)

A positive side effect of getting rid of this assumption is that it would solve the original motivation of https://github.com/kedro-org/kedro/issues/1273, e.g. letting people customize the session run_id, which is particluarly useful when deploying a kedro pipeline to an orchestrator which assumes task independency (like airflow). 

#### Database connection lazy loading

One current issue people are facing when using the session, particularly in interactive mode (but not only), is that database connection & API's calls are instantiated on ``KedroSession.create``, which means that you cannot run your pipeline if another pipeline is not correcly instantiated. I think this would simplify debugging and ease of use if this issue was tackled https://github.com/kedro-org/kedro/issues/2829. 

In general, we should be able to run a pipeline from a session even if other pipelines are broken (no import error, invalid data connection...).  but this is a hard problem and not the prioritary one. 

#### Create a public API for consistent

The lack of public API makes code supporting above use cases become obsolete very fast as kedro versions are changing, because it relies ont a lot of private methods. This severly degrades the developer experience for kedro users :
- beginners are very confused because they find very different examples across the web (https://github.com/kedro-org/kedro/discussions/2134,https://www.youtube.com/watch?v=t9kQ-39qK_g, https://medium.com/indiciumtech/how-to-build-models-as-products-using-mlops-part-4-deploying-models-in-apis-with-kedro-fast-api-15f13e67c080 ...) , and there are hard to adapt to make them work. ChatGPT itself is quite confused between versions and mixes up 0.16/0.17/0.18 examples
- seasoned developers deploying kedro in production increase tech debt and maintenance cost, and this can be a hindrance to kedro adoptions
- plugins developers need to tests in details their plugins and to keep them very up to date - unfortunately this is not something we can reasonably expect for unpaid open source developers, and public plugins for serving are obsolete (https://pypi.org/project/kedro-fast-api/, https://github.com/Galileo-Galilei/kedro-serving), and can add even more confusion for users. 

There is a long standing issue to help make consistent access to kedro objects for plugin developers  https://github.com/kedro-org/kedro/issues/779, and creating a public API to support such use cases is a good step forward.

**Comment by astrojuanlu** [comment]
> This issue, among others, was mentioned on 2024-02-14 tech design, in which @Galileo-Galilei and @takikadiri showed kedro-boot https://github.com/takikadiri/kedro-boot/ there was agreement that implementing this, plus making the `Session` re-entrant and more lightweight #2182 #2879, would be good.

_Originally posted by @astrojuanlu in https://github.com/kedro-org/kedro/issues/2169#issuecomment-1945946338_

**Comment by lordsoffallen** [comment]
This is very much needed, I would like to see seamless serving for kedro pipelines.

Here is how i did to overcome current workarounds to make the pipeline run work multiple times while injecting data and getting the output:

```python
from typing import Iterable, Any
from kedro.framework.project import pipelines
from kedro.framework.session import KedroSession
from kedro.runner import AbstractRunner, SequentialRunner
from kedro.io.core import generate_timestamp


def get_pipeline(
    pipeline_name: str | None = None,
    tags: Iterable[str] | None = None,
    node_names: Iterable[str] | None = None,
    from_nodes: Iterable[str] | None = None,
    to_nodes: Iterable[str] | None = None,
    from_inputs: Iterable[str] | None = None,
    to_outputs: Iterable[str] | None = None,
    namespace: str | None = None,
):
    name = pipeline_name or "__default__"

    try:
        pipeline = pipelines[name]
    except KeyError as exc:
        raise ValueError(
            f"Failed to find the pipeline named '{name}'. "
            f"It needs to be generated and returned "
            f"by the 'register_pipelines' function."
        ) from exc

    filtered_pipeline = pipeline.filter(
        tags=tags,
        from_nodes=from_nodes,
        to_nodes=to_nodes,
        node_names=node_names,
        from_inputs=from_inputs,
        to_outputs=to_outputs,
        node_namespace=namespace,
    )

    return filtered_pipeline


def get_version(request_id: str) -> str:
    return f"{request_id}/{generate_timestamp()}"


class KedroPipeline:
    def __init__(
        self,
        session: KedroSession,
        runner: AbstractRunner | None = None,
        pipeline_name: str | None = None,
        tags: Iterable[str] | None = None,
        node_names: Iterable[str] | None = None,
        from_nodes: Iterable[str] | None = None,
        to_nodes: Iterable[str] | None = None,
        from_inputs: Iterable[str] | None = None,
        to_outputs: Iterable[str] | None = None,
        namespace: str | None = None,
        load_versions: dict[str, str] | None = None
    ):
        self.session = session
        self.runner = runner or SequentialRunner()
        if not isinstance(self.runner, AbstractRunner):
            raise ValueError(
                "KedroPipeline expect an instance of Runner instead of a class."
                "Have you forgotten the `()` at the end of the statement?"
            )

        self.pipeline = get_pipeline(
            pipeline_name=pipeline_name,
            tags=tags,
            node_names=node_names,
            from_nodes=from_nodes,
            to_nodes=to_nodes,
            from_inputs=from_inputs,
            to_outputs=to_outputs,
            namespace=namespace
        )
        self._params = {
            "tags": tags,
            "from_nodes": from_nodes,
            "to_nodes": to_nodes,
            "node_names": node_names,
            "from_inputs": from_inputs,
            "to_outputs": to_outputs,
            "pipeline_name": pipeline_name,
            "namespace": namespace,
            "load_versions": load_versions,
            "runner": getattr(runner, "__name__", str(runner)),
        }

        self.context = self.on_startup()

    def on_startup(self):
        session_id = self.session.store["session_id"]
        save_version = session_id
        extra_params = self.session.store.get("extra_params") or {}
        context = self.session.load_context()

        self._params.update({
            "session_id": session_id,
            "project_path": self.session._project_path.as_posix(),
            "env": context.env,
            "extra_params": extra_params,
            "save_version": save_version
        })

        return context

    def run(self, request_id: str, new_datasets: dict = None) -> tuple[dict[str, Any], Any]:
        """Runs the pipeline with a runner.

        Raises:
            ValueError: If the named or `__default__` pipeline is not
                defined by `register_pipelines`.
            Exception: Any uncaught exception during the run will be re-raised
                after being passed to ``on_pipeline_error`` hook.
        Returns:
            Any node outputs that cannot be processed by the ``DataCatalog``.
            These are returned in a dictionary, where the keys are defined
            by the node outputs.
        """
        # Dataset versioning
        self.context.config_loader.runtime_params.update(
            {"version": get_version(request_id)}
        )

        catalog = self.context._get_catalog(
            save_version=self._params["save_version"],
            load_versions=self._params["load_versions"],
        )

        # Run the runner
        hook_manager = self.session._hook_manager

        if new_datasets is not None:
            catalog.add_feed_dict(new_datasets, replace=True)

        hook_manager.hook.before_pipeline_run(
            run_params=self._params, pipeline=self.pipeline, catalog=catalog
        )
        try:
            run_result = self.runner.run(
                self.pipeline, catalog, hook_manager, self._params["session_id"]
            )
        except Exception as error:
            hook_manager.hook.on_pipeline_error(
                error=error,
                run_params=self._params,
                pipeline=self.pipeline,
                catalog=catalog,
            )
            raise

        hook_manager.hook.after_pipeline_run(
            run_params=self._params,
            run_result=run_result,
            pipeline=self.pipeline,
            catalog=catalog,
        )

        # Load the final dataset again
        out = catalog.load(list(self.pipeline.outputs())[0])
        return run_result, out
```

And this is the app.py file that does fastapi logic:

```python
from fastapi import FastAPI, APIRouter, Body
from pydantic import BaseModel
from contextlib import asynccontextmanager
from .kedro import KedroPipeline
from kedro.framework.session import KedroSession
from kedro.framework.startup import bootstrap_project
from myproj import PROJECT_PATH
from myproj.pipelines import Datasets as ds


pipelines: dict[str, KedroPipeline] = {}


@asynccontextmanager
async def lifespan(app: FastAPI):   # noqa
    bootstrap_project(PROJECT_PATH)
    session = KedroSession.create(PROJECT_PATH, env="staging")
    session._logger.info("Kedro project %s", session._project_path.name)

    for p in ["serving"]:   # Define pipelines to run
        pipelines[p] = KedroPipeline(session=session, pipeline_name=p)

    yield

    session.close()
    pipelines.clear()


app = FastAPI(lifespan=lifespan)
router_v1 = APIRouter(prefix="/v1")


@app.get("/")
async def home():
    return {"msg": "Hello World"}


class ReqInput(BaseModel):
    request_id: str
    text: str


@router_v1.post("/create_text")
async def create_text(request: ReqInput = Body(...)):

    _, output = pipelines["serving"].run(
        request.request_id, {ds.SERVING_DATASET_TEST: request.text}
    )

    return {
        "request_id": request.request_id,
        "output": output,
    }


app.include_router(router_v1)
```

When application boot up, i create the session and context then we invoke kedro run on each incoming requests and since we have versioning data io will also be fine.

Since catalog relies on runtime_params called version which is the request id, i had to keep it in the `run` method as it resolves the pattern already at this stage. 

Quite somethings to improve here but i thought I share my workaround just in case if people need to do something similar.

**Comment by astrojuanlu** [comment]
Thanks @lordsoffallen . Indeed, this is one of the oldest feature requests and the most upvoted one before we closed https://github.com/kedro-org/kedro/issues/143 . Too many different ways of doing this exist.

I'm curious, did you try kedro-boot? https://github.com/takikadiri/kedro-boot/

**Comment by lordsoffallen** [comment]
@astrojuanlu, I had a look into that but it was a bit complicated for my taste, didn't really understand how i can do what i did here. It had some spec configuration and so on so I found it not easy to understand or do it. 

I really needed something simple, like a class that can run a kedro pipeline and easy way to inject a dataset. Most of this logic was in kedro session class anyways so i had to split into functions and make session run multiple times by making sure i don't load context/session each time.

The only hack i had to was to load the final dataset again to return in server. For now it's gonna be fine so i don't worry about it much. 

I think if we make this out of the box work in kedro, that'd be amazing.

**Comment by astrojuanlu** [comment]
> The only hack i had to was to load the final dataset again to return in server.

Rings a bell... 😄 https://github.com/SajidAlamQB/kedro-genai-starter/commit/56203bea97cf1315027913e6558e32e234c88d84
